src/file_pipe_log/pipe_builder.rs

```diff
@@ -100,141 +102,136 @@ impl<F: FileSystem> DualPipesBuilder<F> {
     /// Scans for all log files under the working directory. The directory will
     /// be created if not exists.
     pub fn scan(&mut self) -> Result<()> {
-        let root_path = Path::new(&self.cfg.dir);
-        if !root_path.exists() {
-            info!("Create raft log directory: {}", root_path.display());
-            fs::create_dir(root_path)?;
-            self.dir_lock = Some(lock_dir(root_path)?);
+        self.scan_dir(self.cfg.dir.clone())?; // scan Main dir
+        if let Some(dir) = self.cfg.auxiliary_dir.as_ref() {
+            let auxiliary_dir = dir.clone();
+            self.scan_dir(auxiliary_dir)?; // scan Auxiliary dir
+        }
+        // Sorts the expected `file_list` according to `file_seq`.
+        self.append_files.sort_by(|a, b| a.seq.cmp(&b.seq));
+        self.rewrite_files.sort_by(|a, b| a.seq.cmp(&b.seq));
+        self.recycled_files.sort_by(|a, b| a.seq.cmp(&b.seq));
+
+        // Validate and clear obsolete metadata and log files.
+        for (queue, files, is_recycled_file) in [
+            (LogQueue::Append, &mut self.append_files, false),
+            (LogQueue::Rewrite, &mut self.rewrite_files, false),
+            (LogQueue::Append, &mut self.recycled_files, true),
+        ] {
+            if files.is_empty() {
+                continue;
+            }
+            // Check the file_list and remove the hole of files.
+            let mut current_seq = files[0].seq;
+            let mut invalid_files = 0_usize;
+            debug_assert!(current_seq > 0);
+            for (i, f) in files.iter().enumerate() {
+                if f.seq > current_seq + (i - invalid_files) as u64 {
+                    warn!(
+                        "Detected a hole when scanning directory, discarding files before file_seq {}.",
+                        f.seq,
+                    );
+                    current_seq = f.seq + 1;
+                    invalid_files = i;
+                } else if f.seq < current_seq {
+                    return Err(Error::InvalidArgument("Duplicate file".to_string()));
+                }
+            }
+            files.drain(..invalid_files);
+            // Try to cleanup stale metadata left by the previous version.
+            if files.is_empty() {
+                continue;
+            }
+            let mut cleared = 0_u64;
```

Comment:
I don't think you fully understand the old code. The new implementation is really bad.